refactor: 공통 컴포넌트 업데이트

sdh20282 opened the pull request:
### location input component 미구현 사항 구현
- /search/location 페이지 구현 및 location input component와 연결
![테스트](https://github.com/user-attachments/assets/74f73bf3-4543-4049-be46-fcdc0b5d073b)

### text input component 업데이트
- text input component에서 현재 문자열 길이 출력 부분 구현
![테스트](https://github.com/user-attachments/assets/8cb00868-d545-4b39-bd6a-d66b68d5cf4e)

보다 자세한 사항은 [노션](https://www.notion.so/Components-d6b476544d404b1dab8bb5459ecdee88) 참조

---
### 논의가 필요해 보이는 사항들

검색어 입력
- 사용자가 입력한 검색어에 따라 검색하는 부분에서
- 직접 구현하는 것보다 es-hangul 같은 라이브러리를 사용하는게 어떤지
- https://github.com/toss/es-hangul

search 페이지 동작 방식/데이터 전달
- 현재 사용자가 유저/지역 등 검색 시 어떤 시나리오로 동작하는지 통일되어 있지 않음
- 추가적으로 search 페이지에서 기존 페이지로 이동 시 데이터 처리 방법에 대해 통일 필요

input 유효성 검사 관련
- 금지 단어나 예외 단어 등 관련 내용 통일 필요

sdh20282 committed:
feat: location search page 구현

sdh20282 committed:
feat: 지역 선택 기능 추가
- 다음 history state가 /search/location일 경우 초기화 x
- 이외의 페이지로 이동할 경우에는 store 초기화

sdh20282 committed:
docs: 다음 state 체크 부분 주석 추가

sdh20282 committed:
feat: text input component에서 현재 text length 출력 부분 구현
- valid, invalid 표시

sdh20282 committed:
refactor: text input component jsdoc 업데이트

sdh20282 committed:
refactor: 테스트 케이스 추가

jae-hwan-kim commented on just-swim-app/src/app/_store/location_prev.ts:
해당 코드는 어떤 기능을 하는 건가요??

sdh20282 replied:
이전에 작성된 코드라서 제가 함부로 삭제할 수 없다고 판단해 남겨둔 코드입니다. 다른 부분들도 마찬가지이지만 prev가 붙은 폴더/파일들은 추후에 정리할 필요가 있어보입니다

jae-hwan-kim replied:
넵

jae-hwan-kim commented on just-swim-app/src/app/_store/location.ts:
```diff
@@ -1,8 +1,30 @@
+          }
+        })
+    }),
+    {
+      name: 'location_store',
+      partialize: (state: any) => ({
+        location: state.location,
+      })
+    }
```
1. 중괄호로 스코프를 나눈 이유가 있으신가요??
2. type 을 any 로 설정하신 이유가 있으신가요??

sdh20282 replied:
1. 객체를 반환해줘야 하기 때문에 소괄호로 한번 감싸준 것입니다
2. 수정 완료했습니다

jae-hwan-kim replied:
아! `const 변수이름 = { name: ..., partialize:... }` 의 형태로 선언을 따로 안해줘도 괜찮은건가요??

sdh20282 replied:
제가 이해를 못한 것 같은데

```
const persistStore = {
  name: 'location_store',
  partialize: (state: State & Action & { location: string | undefined | null; }) => ({
    location: state.location,
  })
}

export const locationStore = create<State & Action>()(
  persist(
    (set) => ({
      location: "",
      setLocation: (location: string) => 
        set(() => {
          return {
            location: location,
          }
        })
    }),
    persistStore
  )
)
```
이렇게 말씀하시는건가요?

jae-hwan-kim replied:
네! 아니면 이런 형태요!
```
export const locationStore = create<State & Action>()(
  persist(
    (set) => ({
      location: "",
      setLocation: (location: string) => 
        set(() => {
          return {
            location: location,
          }
        })
    }),
    locataion :  {
         name: 'location_store',
         partialize: (state: State & Action & { location: string | undefined | null; }) => ({
             location: state.location,
          )
     }
)
```

근데 제가 zustand 사용을 많이 안해봐서, 궁금해서 여쭤봤어요.
유저스토어 예시로 들면, 아래처럼 스토어에 객체 선언할 때, 변수 없이 해도 접근할 때 괜찮은가해서요.


```
// 변수 선언 없는 경우
const useUserStore = create<UserStoreType>((set) => ({
   {
     name: '',
     age: '',
     ...
  },

// 변수 선언한 경우
const useUserStore = create<UserStoreType>((set) => ({
  users: {
     name: '',
     age: '',
  ...
  },

```

sdh20282 replied:
저도 zustand를 많이 사용해보지는 않아서 깊게 알지는 못하지만 그래도 제가 아는 내용을 바탕으로 말씀드리자면

![image](https://github.com/user-attachments/assets/c6b2d4a4-30e7-411a-9b84-ed8e7346678c)
우선 persist 함수가 객체 형태로 인자를 전달받는 것이 아니라 해당 부분과 같이 이름을 통해 입력하는 것은 안되는 것으로 알고 있습니다

![image](https://github.com/user-attachments/assets/4c1a926c-4694-4bfa-b6ee-c83bb05ee2a4)
두번째 경우에는 위의 예시와 같이 State를 어떻게 설계하느냐에 따라 가능할 것 같습니다

jae-hwan-kim replied:
넵! 확인했습니다! 감사합니다!

sdh20282 committed:
refactor: location store type 수정

sdh20282 committed:
refactor: date input common component에서 use props 추가

sdh20282 committed:
refactor: 날짜 선택 모달 ui 수정

sdh20282 committed:
feat: Modal Body에 드래그로 모달 닫기 기능 추가

sdh20282 committed:
feat: 모달 스크롤 방지 기능 추가, 모달 역방향 스크롤 방지 구현

sdh20282 committed:
refactor: 모바일 환경에서 slide switch 클릭 시 하이라이트 제거

sdh20282 committed:
refactor: 모바일 버튼 hilight 방지 코드 버튼 전체에 적용

sdh20282 committed:
fix: light comfirm modal에서 이벤트가 연결되어 있지 않은 문제 수정

sdh20282 committed:
refactor: hideModal props 타입을 HTMLButtonElement -> HTMLElement로 변경

sdh20282 committed:
refactor: time picker 내부 컴포넌트인 verticalScroll을 공통 컴포넌트 verticalSlider로 이름, 위치 변경

sdh20282 committed:
refactor: calendar 로직 변경
- 기존 calendar 컴포넌트에서 진행되던 렌더링을 실제 달력을 사용하는 곳에서 렌더링하도록 변경
- 기존 calendar 컴포넌트 로직을 useCalendar 커스텀 훅으로 이동
- 기존 calendar 컴포넌트 최적화 수행 (렌더링 3번 -> 1번)

sdh20282 committed:
feat: month picker 구현, vertical slider 추상화 및 최적화, time picker 로직 수정

sdh20282 committed:
refactor: time picker 내부 간격 수정

sdh20282 committed:
fix: month picker use client 선언 문제 수정

sdh20282 commented:
24.08.11 추가 구현 사항

### date input 컴포넌트 flag 기능 구현
![테스트](https://github.com/user-attachments/assets/fe5bcd76-e823-468d-946f-6a3a97214900)

### calendar 공통 컴포넌트 커스텀 훅으로 변경
- 달력의 렌더링을 컴포넌트 내부 -> 부모로 위임 (조금 더 다양한 디자인 적용 가능하도록)
- 내부 로직 최적화

### vertical slider 컴포넌트 내부 로직 변경, 최적

### calendar에서 월 선택 구현
![테스트](https://github.com/user-attachments/assets/2963411d-2caa-42ac-b91b-086047ce2d36)

### 모달 드래그 구현
![테스트](https://github.com/user-attachments/assets/8c3602a0-e60a-442a-bc11-c98e6aac0641)
- 현재는 특정 픽셀 이상 드래그할 경우 닫히게 설정
- 추후 보다 자세한 로직 상의 필요

기타 자잘한 부분들 수정

sdh20282 committed:
refactor: date modal에서 월 선택 시 title이 보이지 않도록 수정

sdh20282 committed:
feat: month select modal 구현

sdh20282 committed:
refactor: 주에 대한 정보 상수화

sdh20282 committed:
refactor: day select modal에서 day item 디자인 수정

sdh20282 committed:
refactor: time input, time picker에서 시간 표시 수정 (기존 12 -> 00)

sdh20282 committed:
feat: loop vertical slider 구현, time picker에서 시, 분에 적용 및 month picker에서 월에 적용

sdh20282 committed:
feat: modal에 portal 적용

sdh20282 committed:
fix: vertical slider에서 빠르게 드래그 시 item이 밀려서 선택되던 문제 수정

sdh20282 committed:
fix: Portal에 use client 추가

sdh20282 committed:
refactor: vertical loop slider에서 출력할 리스트 길이 수정
- 빠르게 드래그할 때 빈 영역이 보이는 문제 수정
- vertical slider에서 중앙 영역 선택을 위해 준 margin을 transform으로 변경

sdh20282 commented:
24.08,15 구현 사항


### Modal에 Portal 적용
![image](https://github.com/user-attachments/assets/b3d5a49a-f85f-47f6-9aab-9ecbcae0fafd)

### vertical slider loop 컴포넌트 구현
![테스트](https://github.com/user-attachments/assets/313d42d1-d75c-42ce-95a2-71fc72009640)
- 무한 스크롤 기능 구현

jae-hwan-kim commented on just-swim-app/src/app/_types/typeModal.ts:
크게 상관없는 부분인데, Modal 에 공통적으로 사용되는 interface 를 하나로 모아서 extends 하는 것도 괜찮아 보이는데 어떻게 생각하시나요? 

hideModal 을 예로 들면 이렇게 될거 같아요.

```
// 공통 타입 파일
interface ModalActionsProps {
 hideModal: (event: MouseEvent<HTMLElement>) => void;
}

// 관련 파일
export interface DateModalProps extends ModalActionProps {
  initialDate: string,
  setDate: (color: string) => void,
}
```

sdh20282 replied:
좋습니다. 수정할게요

sdh20282 replied:
수정했습니다!

jae-hwan-kim commented on just-swim-app/src/app/_data/date.ts:
```diff
@@ -0,0 +1,19 @@
+  "토",
+] as const;
+
+export const DAY_KOR_TO_ENG: {[props: string]: any} = {
```
혹시 value 를 any 타입으로 설정하신 이유가 있으신가요??

sdh20282 replied:
음.. 수정을 안했었네요.. 수정하겠습니다!

sdh20282 replied:
수정했습니다

jae-hwan-kim commented:
단순한 궁금증인데, useCallback 을 사용하시는 기준이 따로 있으신가요??

sdh20282 commented:
> 단순한 궁금증인데, useCallback 을 사용하시는 기준이 따로 있으신가요??

의존성 배열에 들어가는 요소들이 고정적이고, 연산이 복잡해질 우려가 있는 함수들에서는 적용하는 편입니다..!

sdh20282 committed:
refactor: modal 관련 속성 묶어서 처리

sdh20282 committed:
fix: data에 선언되어 있던 any 제거

sdh20282 committed:
refactor: WEEK_DAYS_TO_ENG 상수 추가, 해당 리스트로 타입 수정

jae-hwan-kim commented:
> > 단순한 궁금증인데, useCallback 을 사용하시는 기준이 따로 있으신가요??
> 
> 의존성 배열에 들어가는 요소들이 고정적이고, 연산이 복잡해질 우려가 있는 함수들에서는 적용하는 편입니다..!

아!! 감사합니다:)

sdh20282 committed:
refactor: date input 컴포넌트에서 defaultValue를 지정하지 않을 시 오늘로 선택되게 수정

sdh20282 committed:
feat: text input 컴포넌트에 에러 메시지 출력 기능 추가
